File: Algorithm/back_tracing/Permutations II/permuteUnique.cpp
class Solution {
public:
	vector<vector<int>> permuteUnique(vector<int>& nums) {
		vector<int>templist;
		vector<bool>visited(nums.size(), false);
		vector<vector<int>>list;
		sort(nums.begin(), nums.end());
		backtrace(list, templist, nums, visited);
		return list;

	}

	void backtrace(vector<vector<int>>&list, vector<int>templist, vector<int>nums, vector<bool> visited)
	{
		if (templist.size() == nums.size()) list.push_back(templist);
		
		for (int i = 0; i<nums.size(); i++)
		{
			
			if (visited[i]){ continue; }
			
			if (i>0 && nums[i] == nums[i - 1]&&visited[i-1]){ continue; }
			visited[i] = true;
			templist.push_back(nums[i]);
			backtrace(list, templist, nums, visited);
			templist.pop_back();
			visited[i] = false;
		}
	}
};

File: Algorithm/back_tracing/Permutations II/permuteUnique.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        lst = []
        nums.sort()
        visited = [0]*n
        def backtrace(temp):
            if len(temp)==n:
                lst.append(temp)
                return 
            for j in range(n):
                if visited[j] == 1: continue
                if j >0 and nums[j]==nums[j-1] and visited[j-1]==0:continue
                visited[j] = 1 
                backtrace(temp+[nums[j]])
                visited[j] = 0
        backtrace([])
        return lst
